### DISCUSSION TITLE
version 4.1.0

### DISCUSSION THREAD

**Discussion by hvianna** [hvianna/audioMotion-analyzer#51] [Announcements]
### Added:

+ [`colorMode`](README.md#colormode-string) - additional options for coloring analyzer bars: by bar amplitude (level) or bar position (index);
+ [`roundBars`](README.md#roundbars-boolean) - rounded corner analyzer bars;
+ [`isRoundBars`](README.md#isroundbars-boolean-read-only) read-only property;
+ [`trueLeds`](README.md#trueleds-boolean) - your LED bars now more vintage than ever!
+ New `level` property for [registered gradients](README.md#registergradient-name-options-).

### Fixed:

+ Avoid drawing the Y-axis scale and unlit LEDs twice, in 'dual-combined' channel layout.

### Improved:

+ Improved frequency scales labeling (especially for `noteLabels`) - label sizes were slightly reduced in fullscreen;
+ Code cleanup and optimizations.

<hr /><em>This discussion was created from the release <a href='https://github.com/hvianna/audioMotion-analyzer/releases/tag/4.1.0'>version 4.1.0</a>.</em>

**Comment by TF3RDL** [comment]
Nice update, but someone commented on the [post I've made](https://www.reddit.com/r/foobar2000/comments/14u3fuj/if_you_have_stereo_mix_you_can_use/) on r/foobar2000, so I updated [my channel spectrum visualization on CodePen](https://codepen.io/TF3RDL/pen/NWBYeaE) to add something like this:

![default ui channel spectrum](https://github.com/hvianna/audioMotion-analyzer/assets/102759506/f6ff6f2f-d4a4-4aaf-beb4-1e8ef49067f3)

So, I'd like to see the `dual-horizontal` property for `channelLayout` that look like this above because foo_uie_vis_channel_spectrum hasn't been updated for a very long time (no default UI support, nor it is compatible with 64-bit version of foobar2000 v2.0 and newer)

**Comment by hvianna** [comment]
Sure, `dual-horizontal` is in the plans and you'll be able to choose different layouts: low or high frequencies meeting at the center, or both channels from low to high.

I wasn't sure if that would be a much desired feature, so thanks for letting me know.

**Reply by TF3RDL** [reply]
Good to know you're focusing on adding new features, as I'd leave the task of porting audioMotion-analyzer as 64-bit foobar2000 component (#50) to someone else as a [part of recreating the magical foo_fft_analyzer](https://hydrogenaud.io/index.php/topic,124330.0.html).

Also, do you need create a separate repository just for a sliding DFT-based CQT/VQT spectrum analyzer (#30) since the sliding DFT requires AudioWorklets, unlike custom FFT, which don't require AudioWorklet and it can be easily accomplished by taking the output of `getFloatTimeDomainData()`?

**Comment by hvianna** [comment]
I plan to add a callback interface so user functions can provide data for visualization, as it has been asked before (on #1 and #3). This would allow using different custom transforms implemented as separate modules.

**Reply by TF3RDL** [reply]
Ah, interesting. Hopefully, the [CQT/VQT algorithm](https://gist.github.com/TF3RDL/0187c076ccae3f57873c5e8cd17917b7) and the [sliding DFT version](https://gist.github.com/TF3RDL/1810c570bad4f75f77941c981c65723f) is compatible with the proposed callback functions for custom transforms, especially the frequency bands part.

BTW, I think getting the X-axis labels positioned correctly on audio spectrum whose frequency band generation algorithm really depends on FFT size/number of bands/sample rate like [my channel spectrum visualization](https://codepen.io/TF3RDL/pen/NWBYeaE) is more difficult to implement accurately as getting a closed-form equation for these frequency band generation output is very difficult or even impossible to get

**Reply by hvianna** [reply]
Thanks for the gists! I'll have a more detailed look at the code as soon as I have a chance.

**Reply by TF3RDL** [reply]
BTW, while I'm racing for my CodePen projects (interactive mockups of foo_fft_analyzer, foo_cqt_analyzer, and other Wamoc's audio analysis components for foobar2000) against Crossover's [foo_enhanced_spectrum_analyzer](https://hydrogenaud.io/index.php/topic,115912.0.html) in terms of updates, I added FFT bandpower mode to [my project that is originally](https://codepen.io/TF3RDL/pen/poQJwRW) for CQT/VQT visualizations, something like this:
![bandpower of shepard tone](https://github.com/hvianna/audioMotion-analyzer/assets/102759506/c7a23370-fe5d-422f-b88f-be17c7544bdf)
which is a lanzcos-interpolated FFT bandpower reminiscent of foobar2000's default spectrum or even WhiteCap Standalone, which is different from the CQT algorithm like this below:
![vqt of shepard tone](https://github.com/hvianna/audioMotion-analyzer/assets/102759506/a2379e89-64dc-4af7-bfe6-04ee61e4f349)

BTW, even doing an FFT with a non-power of 2 size is faster than calculating a CQT/VQT directly on large number of bands, whereas direct CQT/VQT calculation might be faster than doing a NPoT (acronym for non-power of two) FFT with third-octave or fewer number of bands
